In [31]:
import pandas as pd
import numpy as np
from difflib import SequenceMatcher
import Levenshtein as lev

In [38]:
lev.ratio("Joseph Walden Johnson Jr.", "Joseph Johnson")

0.717948717948718

In [30]:
SequenceMatcher(None, "Joseph Walden Johnson Jr.", "Joseph Johnson").ratio()
"Joseph Johnson" in "Joseph Walden Johnson Jr."

False

In [3]:
data3 = pd.read_csv("killedbypolice.csv")
data2 = pd.read_csv("MPVDatasetDownload.csv")
data1 = pd.read_csv("https://raw.githubusercontent.com/washingtonpost/data-police-shootings/master/fatal-police-shootings-data.csv")

In [4]:
data1.head()

id                name        date   manner_of_death       armed   age  \
0   3          Tim Elliot  2015-01-02              shot         gun  53.0   
1   4    Lewis Lee Lembke  2015-01-02              shot         gun  47.0   
2   5  John Paul Quintero  2015-01-03  shot and Tasered     unarmed  23.0   
3   8     Matthew Hoffman  2015-01-04              shot  toy weapon  32.0   
4   9   Michael Rodriguez  2015-01-04              shot    nail gun  39.0   

  gender race           city state  signs_of_mental_illness threat_level  \
0      M    A        Shelton    WA                     True       attack   
1      M    W          Aloha    OR                    False       attack   
2      M    H        Wichita    KS                    False        other   
3      M    W  San Francisco    CA                     True       attack   
4      M    H          Evans    CO                    False       attack   

          flee  body_camera  
0  Not fleeing        False  
1  Not fleeing        False  
2  Not fleeing        False  
3  Not fleeing        False  
4  Not fleeing        False

In [5]:
data2.head()

Victim's name Victim's age Victim's gender Victim's race  \
0             Saheed Vassell           34            Male         Black   
1      Timothy Orlando Wyatt           41            Male         Black   
2  Joseph Walden Johnson Jr.           24            Male         White   
3               Bobby Hinton           60            Male  Unknown race   
4    Christopher Lamarr Hall           34            Male         Black   

                              URL of image of victim  \
0                                                NaN   
1  http://www.fatalencounters.org/wp-content/uplo...   
2  http://www.fatalencounters.org/wp-content/uplo...   
3                                                NaN   
4                                                NaN   

  Date of Incident (month/day/year) Street Address of Incident       City  \
0                            4/4/18                        NaN    Brookyn   
1                            4/1/18     E 28th Ave & N 15th St      Tampa   
2                            4/1/18            Railroad Avenue  Covington   
3                            4/1/18            Magnolia Street  New Llano   
4                            4/1/18     4100 SW Clyde Park Ave    Wyoming   

  State  Zipcode     ...     Criminal Charges?  \
0    NY      NaN     ...      No known charges   
1    FL  33605.0     ...      No known charges   
2    LA  70433.0     ...      No known charges   
3    LA  71461.0     ...      No known charges   
4    MI  49509.0     ...      No known charges   

  Link to news article or photo of official document  \
0                                                NaN   
1  http://www.wfla.com/news/hillsborough-county/p...   
2  http://www.nola.com/crime/index.ssf/2018/04/st...   
3  http://www.klfy.com/top-stories/deputy-involve...   
4  http://www.mlive.com/news/jackson/index.ssf/20...   

  Symptoms of mental illness?          Unarmed Alleged Weapon (Source: WaPo)  \
0                         Yes          Unclear                    metal pipe   
1                          No  Allegedly Armed                           gun   
2                          No  Allegedly Armed                           gun   
3                          No  Allegedly Armed                           gun   
4                         Yes  Allegedly Armed                           gun   

  Alleged Threat Level (Source: WaPo) Fleeing (Source: WaPo)  \
0                              attack            Not fleeing   
1                              attack                    Car   
2                              attack                    Car   
3                              attack            Not fleeing   
4                              attack            Not fleeing   

  Body Camera (Source: WaPo) WaPo ID (If included in WaPo database)  \
0                         No                                 3577.0   
1                         No                                 3564.0   
2                         No                                 3563.0   
3                         No                                 3562.0   
4                         No                                 3566.0   

  Unnamed: 24  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN  

[5 rows x 25 columns]

In [6]:
data3.head()

age            date gender manner of death                    name race  \
0   25  April 21, 2018      M            shot        Terrance Carlton  NaN   
1   48  April 20, 2018      M            shot            James Bauduy  NaN   
2   40  April 19, 2018      M            shot         Charles Whitley  NaN   
3   45  April 18, 2018      M            shot  William Ray Simcoe Sr.    W   
4   48  April 18, 2018      M            shot    David Gino Teneyuque  NaN   

  state  month  year  
0    TN  April  2018  
1    FL  April  2018  
2    KY  April  2018  
3    MO  April  2018  
4    MI  April  2018

In [7]:
data1names = list(data1["name"])
data2names = list(data2["Victim's name"])
data3names = list(data3["name"])

### Dates After 2015

In [8]:
dates2 = list(data2["Date of Incident (month/day/year)"])
months2 = [int(date[:date.find("/")]) for date in dates2]
years2 = [int("20"+date[date.rfind("/")+1:]) for date in dates2]
data2["month"] = np.array(months2)
data2["year"] = np.array(years2)
past20152 = data2.loc[((data2["year"]>=2015) & (data2["year"]<2018)) | ((data2["month"]<4)&(data2["year"]==2018))]["Victim's name"]

In [9]:
dates3 = list(data3["date"])
months3 = [date.split()[0] for date in dates3]
years3 = [int(date.split()[2]) for date in dates3]
data3["month"] = np.array(months3)
data3["year"] = np.array(years3)
past20153 = data3.loc[((data3["year"]>=2015) & (data3["year"]<2018)) | (((data3["month"] == "January") | (data3["month"] == "February") | (data3["month"]== "March")) & (data3["year"]==2018))]["name"]

In [12]:
wsp = data1names
mpv = list(past20152)
kbp = list(past20153)

In [55]:
# cleaning data
newmpv = []
for name in mpv:
    if name != "Name withheld by police":
        newmpv.append(name)
mpv = newmpv
newwsp = []
for name in wsp:
    if name.lower() != "tk tk":
        newwsp.append(name)
wsp = newwsp

In [50]:
print("Number of names in the Washington Post dataset:", "\t", len(wsp))
print("Number of names in the Mapping Police Violence dataset:", "\t", len(mpv))
print("Number of names in the Killed By Police dataset:", "\t", len(kbp))

Number of names in the Washington Post dataset: 	 3208
Number of names in the Mapping Police Violence dataset: 	 3702
Number of names in the Killed By Police dataset: 	 3148


## Common Names Across All 3 Datasets

In [58]:
list12 = []
for name1 in wsp:
    for name2 in mpv:
        if lev.ratio(name1, name2) > 0.7:
            list12.append(name2)
            break

In [59]:
ls = []
for name in list12:
    for name3 in kbp:
        if lev.ratio(name, name3) > 0.7:
            ls.append(name3)
            break
len(ls)

2739

## Common Names Across Killed By Police and Washington Post

In [56]:
list12 = []
list12missing = []
for name1 in wsp:
    found = False
    for name2 in kbp:
        if lev.ratio(name1, name2) > 0.7:
            list12.append(name1)
            found = True
            break
    if not found:
        list12missing.append(name1)
len(list12)

2786

### Missing names from Killed by police that are in washington post

In [57]:
list12missing

['Omarr Julian Maximillian Jackson',
 'Edward Donnell Bright',
 'Dewayne Deshawn Ward',
 'Desmond Luster',
 'Howard Brent Means Jr.',
 'Alexander Phillip Long',
 'Charly Leundeu Keunang',
 'Hung Trieu',
 'Mya Hall',
 'Frank Shephard',
 'Billy Joe Patrick',
 'Nadir Soofi',
 'Dennis  Fiel',
 'James Strong',
 'Lorenzo Garza',
 'Jesus Quezada Gomez',
 'Raymond Peralta-Lantigua',
 'Victor Emanuel Larosa',
 'Paul Castaway',
 'Adam Schneider',
 'Kyle Lambrose',
 'Jose Ramon Damiani Jr',
 'Angelo Delano Perry',
 'India Kager',
 'Bruce Santino',
 'Rory Lynn Gunderman',
 'Tina Money',
 'Norma Angelica Guzman',
 'Leslie Portis',
 'Martin  Ryans Jr.',
 'Roger D.  Hall',
 'Ryan Christopher Rodriguez',
 'Tony Berry',
 'Shane Whitehead',
 'Tashfeen Malik',
 'Syed Farook',
 'Bettie Jones',
 'Quintonio LeGrier',
 'Gayneaux Paul Trahan',
 'Keith Childress',
 'Lonnie Haskell Powers',
 'Clarence Layfield',
 'Shawn Henry Allen Dickens',
 'Jose Quintanilla',
 'Oscar Reyes',
 'Che Taylor',
 'Kisha Michael',


## Common Names Across Mapping Police Violence and Washington Post

In [45]:
list13 = []
for name1 in wsp:
    for name2 in mpv:
        if lev.ratio(name1, name2) > 0.7:
            list13.append(name2)
            break
len(list13)

3120

## Common Names Across Killed By Police and Mapping Police Violence

In [46]:
list23 = []
for name1 in kbp:
    for name2 in mpv:
        if lev.ratio(name1, name2) > 0.7:
            list23.append(name2)
            break
len(list23)

2983

## Other Tries at Analysis

In [34]:
# creating columns for first and last name in each data set
firstlast = []
for name in data2names:
    firstname = name[0:name.find(" ")]
    lastname = name[name.rfind(" ")+1: ]
    if len(lastname) <= 3:
        name = name[0:name.rfind(" ")]
        lastname = name[name.rfind(" "):]
    firstlast.append(firstname + " " + lastname)
data2["first and last name"] = firstlast
data2

Victim's name Victim's age Victim's gender  \
0                   Saheed Vassell           34            Male   
1            Timothy Orlando Wyatt           41            Male   
2        Joseph Walden Johnson Jr.           24            Male   
3                     Bobby Hinton           60            Male   
4          Christopher Lamarr Hall           34            Male   
5              Cresencio Rodriguez           33            Male   
6                    Brian Bellamy           41            Male   
7                Luis Yair Alvarez           22            Male   
8          Name withheld by police      Unknown            Male   
9                     John Wussler           58            Male   
10                 Jason Alan Birt           27            Male   
11                  Nicolas Sedano           21            Male   
12               Edward Van McCrae           60            Male   
13            Timothy Breckenridge           42            Male   
14                  Susan LoTempio           64          Female   
15                  Dishon McBride           19            Male   
16          Marco Antonio Saavedra           44            Male   
17                    Rodney Toler           39            Male   
18         Christopher M. Gatewood           39            Male   
19         James Kenneth Decoursey           34            Male   
20                    Jesse Kilgus           51            Male   
21             Aaron Wagdy Ibrahem           26            Male   
22                  Robert B. Shaw           29            Male   
23          William John Dominguez           31            Male   
24                George M. Pappas           61            Male   
25         Roderic Bernard Cameron           44            Male   
26    Gerald "Ricky" Richard Johns           64            Male   
27             Jose Aaron Gonzalez           44            Male   
28              Juan Carlos Romero           41            Male   
29          Isaac Michael Scroggie           24            Male   
...                            ...          ...             ...   
5964               Angella Falconi           24          Female   
5965               Cedrick Chatman           17            Male   
5966               Vincent Jimenez           35            Male   
5967           Jimmy L. Hamlin Jr.           44            Male   
5968                Daniel Brawley           29            Male   
5969      Spencer Rollins Mims III           55            Male   
5970               Jeremy Rucinski           38            Male   
5971             Daniel Autenrieth           31            Male   
5972               Sonny Archuleta           33            Male   
5973                Bernard Rowley           55            Male   
5974               Theodore Keiper           58            Male   
5975                 David Compton           27            Male   
5976               Chuckie Stowers           39            Male   
5977                     Joel Byne           52            Male   
5978           Joseph Blake Powell           21            Male   
5979               Barry Cloninger           51            Male   
5980         Xavier Tyrell Johnson           31            Male   
5981                Yolanda Thomas           34          Female   
5982             Seth W. O'Donnell           26            Male   
5983                 Darrell Banks           47            Male   
5984             Hunter Jacob Todd           20            Male   
5985           Amber Nashay Carter           21          Female   
5986                 Peter Jourdan           37            Male   
5987                Kenneth Morrow           52            Male   
5988                  Abel Gurrola           26            Male   
5989             Andrew L. Closson           21            Male   
5990                   Mark Chavez           49            Male   
5991                 Andrew Layton           26            Male   
5

In [40]:
firstlast = []
for name in data1names:
    firstname = name[0:name.find(" ")]
    lastname = name[name.rfind(" ")+1: ]
    if len(lastname) <= 3:
        name = name[0:name.rfind(" ")]
        lastname = name[name.rfind(" "):]
    firstlast.append(firstname + " " + lastname)
data1["first and last name"] = firstlast
data1

id                              name        date   manner_of_death  \
0        3                        Tim Elliot  2015-01-02              shot   
1        4                  Lewis Lee Lembke  2015-01-02              shot   
2        5                John Paul Quintero  2015-01-03  shot and Tasered   
3        8                   Matthew Hoffman  2015-01-04              shot   
4        9                 Michael Rodriguez  2015-01-04              shot   
5       11                 Kenneth Joe Brown  2015-01-04              shot   
6       13               Kenneth Arnold Buck  2015-01-05              shot   
7       15                     Brock Nichols  2015-01-06              shot   
8       16                     Autumn Steele  2015-01-06              shot   
9       17                   Leslie Sapp III  2015-01-06              shot   
10      19                    Patrick Wetter  2015-01-06  shot and Tasered   
11      21                         Ron Sneed  2015-01-07              shot   
12      22    Hashim Hanif Ibn Abdul-Rasheed  2015-01-07              shot   
13      25            Nicholas Ryan Brickman  2015-01-07              shot   
14      27  Omarr Julian Maximillian Jackson  2015-01-07              shot   
15      29                     Loren Simpson  2015-01-08              shot   
16      32               James Dudley Barker  2015-01-08              shot   
17      36               Artago Damon Howard  2015-01-08              shot   
18      37                      Thomas Hamby  2015-01-08              shot   
19      38                     Jimmy Foreman  2015-01-09              shot   
20     325                     Andy Martinez  2015-01-09              shot   
21      42                       Tommy Smith  2015-01-11              shot   
22      43                    Brian  Barbosa  2015-01-11              shot   
23      45                 Salvador Figueroa  2015-01-11  shot and Tasered   
24      46               John Edward O'Keefe  2015-01-13              shot   
25      48                 Richard McClendon  2015-01-13              shot   
26      49                     Marcus Golden  2015-01-14              shot   
27      50                    Michael Goebel  2015-01-14              shot   
28      51                      Mario Jordan  2015-01-14              shot   
29      52                  Talbot Schroeder  2015-01-14              shot   
...    ...                               ...         ...               ...   
3252  3600             Shawn Michael Hubbard  2018-04-10              shot   
3253  3601                       Mark Powell  2018-04-10              shot   
3254  3602                    Grechario Mack  2018-04-10              shot   
3255  3603            Antonino Thomas Gordon  2018-04-10              shot   
3256  3605                     Betty Lemoine  2018-04-11              shot   
3257  3606                      Kenneth Ross  2018-04-11              shot   
3258  3607                   Keith Alan Kent  2018-04-11              shot   
3259  3613              Chad Eric Montgomery  2018-04-11              shot   
3260  3608                   Dashaun Shepard  2018-04-12              shot   
3261  3609                    Benjamin Evans  2018-04-12              shot   
3262  3610                   Rumondale Jones  2018-04-12              shot   
3263  3611                     Steven Brooks  2018-04-12              shot   
3264  3614                             TK TK  2018-04-12              shot   
3265  3612           Kenneth Warren Resendez  2018-04-13              shot   
3266  3621                Bruce Arnold Allee  2018-04-13              shot   
3267  3615                    Richard Rangel  2018-04-14              shot   
3268  3617                       Jose Pietri  2018-04-14              shot   
3269  3618               Andre Lavance Rippy  2018-04-14              shot   
3270  3619                   Jessie Thedford  2018-04-14              shot   
3271  3620                   E

In [41]:
firstlast = []
for name in data3names:
    firstname = name[0:name.find(" ")]
    lastname = name[name.rfind(" ")+1: ]
    if len(lastname) <= 3:
        name = name[0:name.rfind(" ")]
        lastname = name[name.rfind(" "):]
    firstlast.append(firstname + " " + lastname)
data3["first and last name"] = firstlast
print(len(data3))
data3

4414


age                 date gender manner of death  \
0      48       April 20, 2018      M            shot   
1      40       April 19, 2018      M            shot   
2      45       April 18, 2018      M            shot   
3      48       April 18, 2018      M            shot   
4      32       April 18, 2018      M            shot   
5      32       April 18, 2018      M            shot   
6      25       April 17, 2018      M            shot   
7      39       April 17, 2018      M            shot   
8      41       April 17, 2018      M            shot   
9      59       April 15, 2018      M            shot   
10     36       April 15, 2018      M            shot   
11     32       April 14, 2018      M            shot   
12     55       April 14, 2018      M            shot   
13     44       April 14, 2018      M            shot   
14     39       April 14, 2018      M            shot   
15     17       April 13, 2018      M         vehicle   
16     31       April 13, 2018      M            shot   
17     34       April 13, 2018      M            shot   
18     39       April 12, 2018      M            shot   
19     45       April 12, 2018      M            shot   
20     23       April 12, 2018      M            shot   
21     39       April 12, 2018      M            shot   
22     61       April 11, 2018      M            shot   
23     71       April 11, 2018      F            shot   
24     25       April 11, 2018      M            shot   
25     28       April 10, 2018      M            shot   
26     30       April 10, 2018      M            shot   
27     39       April 10, 2018      M            shot   
28     44       April 10, 2018      M            shot   
29     29        April 9, 2018      M            shot   
...   ...                  ...    ...             ...   
4384   45    October 13, 2014       F            shot   
4385   32    October 13, 2013       M            shot   
4386   30    October 13, 2013       M            shot   
4387   51    October 13, 2013       M            shot   
4388   60    October 12, 2013       M  physical force   
4389   41    October 12, 2013       M  physical force   
4390   21  October 11, 2013 \t      M            shot   
4391   40  October 11, 2013 \t      M             NaN   
4392   55   October 9, 2013 \t      M            shot   
4393   49   October 7, 2013 \t      M            shot   
4394   39    October 6, 2013\t      M             NaN   
4395   24    October 5, 2013\t      M  physical force   
4396   25    October 5, 2013\t      M             NaN   
4397   21    October 5, 2013\t      M            shot   
4398   26   October 5, 2013 \t      M            shot   
4399   43   October 4, 2013 \t      M            shot   
4400   49   October 4, 2013 \t      M         vehicle   
4401   34   October 3, 2013 \t      F             NaN   
4402   39    October 2, 2013\t      M            shot   
4403   45    October 1, 2013\t      F            shot   
4404   37  September 30, 2013       M            shot   
4405   19  September 27, 2013       M            shot   
4406   21  September 27, 2013       M             NaN   
4407   57  September 27, 2013       M            shot   
4408   39  September 26, 2013       M            shot   
4409   34  September 26, 2013       M            shot   
4410   22  September 25, 2013       M            shot   
4411   27  September 25, 2013       M             NaN   
4412   32  September 25, 2013       M            shot   
4413   21  September 24, 2013       M            shot   

                         name race state first and last name  
0                James Bauduy  NaN    FL        James Bauduy  
1             Charles Whitley  NaN    KY     Charles Whitley  
2      William Ray Simcoe Sr.    W    MO     William  Simcoe  
3        David Gino Teneyuque  NaN    MI     David Teneyuque  
4           Delorean Pikyavit    I    UT   Delorean Pikyavit  
5                Justin Oakes    W    GA        Justin Oakes  
6                Sanchez Lowe 

In [36]:
data3names = np.array(data3["name"])
print(len(data3names))
print(data3names)

4414
['James Bauduy' 'Charles Whitley' 'William Ray Simcoe Sr.' ...,
 'Eric Poore' 'Erick Balint' 'Connor Bishop Zion']


In [35]:
# name1 = data1["first and last name"]
# name2 = data2["first and last name"]
# count12 = 0
# for n1 in name1:
#     for n2 in name2:
#         if similar(n1, n2) > 0.4:
#             count12 += 1
#             break
            
# print(count12)

3282


In [37]:
name1 = data1["first and last name"]
name2 = data2["first and last name"]
count12 = 0
for n1 in name1:
    for n2 in name2:
        if n1 in n2:
            count12 += 1
            break
            
print(count12)

2852


name1 with name2 == washington post vs mpv data == 2852 matches

In [50]:
# def similar(data1names, data2names):
#     return SequenceMatcher(None, data1names,data2names).ratio()

# count = 0
# for name2 in data1names:
#     for name1 in data2names:
#         if similar(name1, name2) > 0.4:
#             count += 1
#             break
            
# print(count)

In [43]:
name3 = data3["first and last name"]
count23 = 0
for n2 in name2:
    for n3 in name3:
        if n2 in n3:
            count23 += 1
            break
            
print(count23)

# comparing MPV (data2) with KilledByPolice (data3) == 3878 matches

3878


In [47]:
name1 = data1["first and last name"]
name3 = data3["first and last name"]
countall3 = 0
both23 = []
for n2 in name2:
    for n3 in name3:
        if n2 in n3:
            both23.append(n2)
            break
print(len(both23))
# put first two for loops in a list (descending order) 

3878


In [51]:
all3 = []
for both in both23:
    for n1 in name1:
        if n1 in both:
            all3.append(n1)
            break
print(len(all3))
print(all3)

2305
['Saheed Vassell', 'Timothy Wyatt', 'Bobby Hinton', 'Christopher Hall', 'Cresencio Rodriguez', 'Brian Bellamy', 'Luis Alvarez', 'John Wussler', 'Jason Birt', 'Nicolas Sedano', 'Edward McCrae', 'Timothy Breckenridge', 'Marco Saavedra', 'Rodney Toler', 'Christopher Gatewood', 'James Decoursey', 'Jesse Kilgus', 'Aaron Ibrahem', 'Robert Shaw', 'Gerald Johns', 'Jose Gonzalez', 'Juan Romero', 'Isaac Scroggie', 'Evan Bashir', 'Nathan Castle', 'Kaitlin Demeo', 'Zander Clark', 'Nikolai Yakunin', 'Michael McGinnis', 'Kevin Robles', 'Daniel Young', 'Kevin Hall', 'Linus Phillip', 'Santiago Calderon', 'Danny Thomas', 'Duane Preciado', 'Jason Raffaeli', 'Jehad d', 'Mark Lunn', 'Chance Haegele', 'Kent Pittman', 'Scott Mielentz', 'Joel Pflum', 'Steven Nguyen', 'Michael Holliman', 'Jermaine Massey', 'Manuel Borrego', 'Stephon Clark', 'William Langfitt', 'James Vaughn', 'Bobby Lovin', 'Cameron Hall', 'Shermichael Ezeff', 'Ryan Chapman', 'Bryan Liles', 'Decynthia Clements', 'Michael Ward', 'Jontell 

"which data set is most reliable or "inclusive"? Why are certain names caught across all three and others not?"

MPV : had the most inputted names (data2)
    2305 / 5994 = 38%

KilledBy Police : had second most names (data3)
    2305 / 4414 = 52%

WashingtonPost : had least names (data1)
    2305 / 3282 = 70%